evutil.c

```diff
@@ -480,13 +480,37 @@ evutil_make_listen_socket_reuseable(evutil_socket_t sock)
 int
 evutil_make_listen_socket_reuseable_port(evutil_socket_t sock)
 {
-#if defined __linux__ && defined(SO_REUSEPORT)
-	int one = 1;
-	/* REUSEPORT on Linux 3.9+ means, "Multiple servers (processes or
-	 * threads) can bind to the same port if they each set the option. */
-	return setsockopt(sock, SOL_SOCKET, SO_REUSEPORT, (void*) &one,
-	    (ev_socklen_t)sizeof(one));
+#if defined(__FreeBSD__) && __FreeBSD__ >= 12 && defined(SO_REUSEPORT_LB)
+	/* FreeBSD 12 introduced a new socket option named SO_REUSEPORT_LB
+	 * with the capability of load balancing, it's the equivalent of
+	 * the SO_REUSEPORTs on Linux and DragonFlyBSD. */
+	int enabled = 1;
+	return setsockopt(sock, SOL_SOCKET, SO_REUSEPORT_LB, (void*)&enabled,
+	    (ev_socklen_t)sizeof(enabled));
+#elif (defined(__linux__) || defined(__DragonFly__) || defined(__sun)) && defined(SO_REUSEPORT)
+	int enabled = 1;
+	/* SO_REUSEPORT on Linux 3.9+ means, "Multiple servers (processes or
+	 * threads) can bind to the same port if they each set the option".
+	 * In addition, the SO_REUSEPORT implementation distributes connections
+	 * or datagrams evenly across all of the threads (or processes).
+	 *
+	 * DragonFlyBSD 3.6.0 extended SO_REUSEPORT to distribute workload to
+	 * available sockets, which make it the same as Linux's SO_REUSEPORT.
+	 *
+	 * Solaris 11 supported SO_REUSEPORT, but it's implemented only for
```

Comment:
Great comment, thanks! Shouldn't be check for dragonflybsd/solaris versions then?

Reply:
Well, we can check for dragonflybsd version via macro `__DragonFly_version`, but v3.6.0 (released in 2013) is pretty old, it seems needless. Just like we don't bother to check for linux 3.9.

As for Solaris 11.4, it would be nice to check the version via pre-defined compiler macros of OS, but unfortunately, it's undoable. If we want to detect the kernel version, we may use some tricks.

Reply:
Let's check the presence of the socket option [`SO_FLOW_NAME`](https://docs.oracle.com/cd/E88353_01/html/E37843/setsockopt-3c.html) that was first introduced in solaris 11.4 to make sure we're on solaris 11.4.

Reply:
Let's add at least a comment for the SO_FLOW_NAME option check

P.S. this whole story is not ideal, since we are checking on the compile time, while it could be compiled on newer machine and used on an old one, I guess we could use some sysctl interface for this, and dynamically check, but, let's don't bother about this but document this behaviour in the header file.

Reply:
Comment updated.

But I don't think we need to explicitly document the "it could be compiled on newer machine and used on an old one", this kind of thing goes without saying, and it could happen to all projects that use pre-defined compiler macros.